### PR TITLE
refactor(graph): build GraphDb directly from aptu-coder-core structs; delete text-format parsers

### LINKED ISSUE #1443: refactor(graph): build GraphDb directly from aptu-coder-core structured output; delete text-format parser
## Summary

`builder.rs` re-parses the `<ast_context>` and `<call_graph>` strings that
`ast_context.rs` already produced from `aptu-coder-core` typed structs. The
round-trip is:

```
aptu-coder-core FunctionInfo -> compact_signature() string -> parser -> petgraph Node
```

The structured data (`SemanticAnalysis`, `CallGraph`) is available at the
`analyze_file()` call site in `ast_context.rs` and is discarded. The text
format is not a protocol boundary: `aptu-core` already depends on
`aptu-coder-core` (`Cargo.toml:37`, feature `ast-context`). The parser is an
artifact of the no-second-parse-pass constraint from #1408, which is satisfied
either way: calling `analyze_file()` once and building the graph from its
output is not a second parse pass.

### `graph/builder.rs`

Replace `parse_ast_context_string` and `parse_call_graph_string` with a new
function:

```rust
pub fn build_from_analysis(semantic: &SemanticAnalysis, call_graph: &CallGraph) -> GraphDb
```

Builds `Node::File`, `Node::Function` (and `Struct`, `Enum`, `Trait`, `Impl`
where the analysis provides them), `Edge::Contains`, `Edge::Calls`,
`Edge::Imports` directly from typed fields. No string parsing. No format
contract. No silent degradation on format drift.

Delete `parse_ast_context_string`, `parse_call_graph_string`, and all their
tests. Delete the `# Format contract` doc comment and the
`test_parse_ast_context_skips_unrecognised_lines` canary test.

### `ast_context.rs`

After the existing `analyze_file()` call, pass `&analysis.semantic` and the
`CallGraph` to `build_from_analysis`. Return (or make available to
`build_ctx_graph`) the resulting `GraphDb` alongside the existing text strings.
The text strings continue to be produced unchanged for the LLM prompt.

### `review_context.rs`

`build_ctx_graph` receives the `GraphDb` built in `ast_context.rs` instead of
calling `cache::load_or_build` with text strings. `cache::load_or_build` still
handles disk persistence (cache hit path unchanged).

Fix #1439 item 1 in the same PR: derive `modified_symbols` from `pr.files`
changed hunks (filenames + symbols extracted from the diff) instead of
`extract_function_names_from_ast`.

## Acceptance criteria

- [ ] `parse_ast_context_string` and `parse_call_graph_string` deleted; no text parsing in `graph/builder.rs`
- [ ] `build_from_analysis` builds a correct `GraphDb` from `SemanticAnalysis` and `CallGraph` without any string intermediate
- [ ] Text strings (`<ast_context>`, `<call_graph>`) still produced for the LLM prompt; no regression on prompt content
- [ ] Modified symbols derived from `pr.files` diff, not full graph node set (closes #1439 item 1)
- [ ] Existing `cache::load_or_build` disk persistence path unchanged
- [ ] `cargo test` passes; `cargo clippy -- -D warnings` clean
- [ ] `cargo check -p aptu-core --target wasm32-unknown-unknown --no-default-features` passes

## Closes / relates

- Closes #1439 items 1 and 3
- Relates to #1440 (`cache.rs` unchanged by this PR)
- Relates to #1441 (`name_to_idx` finding obsoleted; noted in amended #1441)
- Relates to aptu-coder#1362 (C-hybrid; schema convergence deferred to that issue)

### PR DESCRIPTION
**Pull request by clouatre**
## Summary

Replace the text-format parsers (`parse_ast_context_string`, `parse_call_graph_string`) in `graph/builder.rs` with a new function `build_from_analysis(path, semantic, call_graph) -> GraphDb` that builds the structural graph directly from typed `aptu-coder-core` structs (`SemanticAnalysis`, `CallGraph`). Also fixes the blast-radius seeding bug from #1439 item 1 by deriving `modified_symbols` from PR diff hunks instead of all graph nodes.

### `graph/builder.rs`

- Delete `parse_ast_context_string`, `parse_call_graph_string`, their `# Format contract` doc comment, and all 7 tests including the `test_parse_ast_context_skips_unrecognised_lines` canary
- Add `pub fn build_from_analysis(path: &str, semantic: &SemanticAnalysis, call_graph: &CallGraph) -> GraphDb` gated on `#[cfg(all(feature = "ast-context", feature = "graph"))]`
  - Emits `Node::File` + `Node::Function` + `Edge::Contains` for each function in `semantic.functions`
  - Emits `Node::Module` + `Edge::Imports` for each entry in `semantic.imports`
  - Emits `Edge::Calls` from `call_graph.callers`, filtering `neighbor_name == "<reference>"` pseudo-edges and `is_impl_trait == true` edges
  - Defaults `visibility` to `"private"` (no visibility field in `FunctionInfo`)
  - Normalizes `CallEdge.path` to `""` for anonymous callers

### `ast_context.rs`

- Add `AstContextOutput` struct (replaces the raw `String` return from `build_ast_context` / `build_ast_context_sync`): carries `text: String` for prompt injection and, when both `ast-context` and `graph` features are enabled, `graph: GraphDb` built from the same analysis pass
- `build_ast_context_sync` accumulates `(PathBuf, SemanticAnalysis)` pairs and `ImplTraitInfo` during its first (and only) `analyze_file` pass
- After the loop: calls `CallGraph::build_from_results(pairs, &impl_traits, false)` once, then builds per-file `GraphDb` instances from the retained pairs (no second `analyze_file` call), merging them into a single `GraphDb` stored on `AstContextOutput`
- Existing tests updated to access `.text` instead of the raw string return

### `graph/cache.rs`

- `load_or_build` input signature changed from `(owner, repo, sha, ast: &str, call_graph: &str, cfg)` to `(owner, repo, sha, graph: GraphDb, cfg)`; return type `(GraphDb, bool)` is unchanged
- Cache-miss path persists the provided `GraphDb`; cache-hit path returns cached and drops the provided graph
- WASM variant: returns provided `GraphDb` with `cache_hit = false`
- `FORMAT_VERSION`, `strip_modifies_edges`, `encode_graph`, `decode_graph` unchanged

### `review_context.rs`

- Add `SYMBOL_RE`: a `LazyLock<Regex>` static that matches Rust declaration syntax (`fn`/`async fn`, `struct`, `enum`, `trait`, `impl`), stripping visibility prefixes including `pub(crate)` and `pub(super)`; gated on `#[cfg(all(feature = "ast-context", feature = "graph"))]`
- `build_ctx_ast` and `build_ctx_graph` each split into a `#[cfg(feature = "ast-context")]` variant and a stub `#[cfg(not(feature = "ast-context"))]` variant for WASM compatibility
- `build_ctx_graph` now accepts `&AstContextOutput` and passes its `graph` field to `cache::load_or_build` (previously accepted `ast_context: &str` and `call_graph: &str`)
- Replace `extract_function_names_from_ast` (returned ALL graph nodes) with `derive_modified_symbols` that parses `pr.files[].patch` diff lines for function/struct/enum/trait/impl declarations; handles `patch = None` (empty), `patch_truncated` (partial), and renamed files
- Add 8 unit tests for `derive_modified_symbols` covering: `patch = None`, basic hunk extraction (fn/struct/enum/impl), truncated patch, renamed file, `async fn`, `pub(crate)`/`pub(super)` visibility, generic type parameters, and tuple/unit structs

### `graph/mod.rs`

- Updated module doc to describe typed struct-based construction
- Re-export updated to expose `build_from_analysis`

## Known trade-off

The typed `CallGraph` is built from the PR's changed files only. The previous text-based approach used `analyze_focused` (a repo-wide walk per function), which included callers from unchanged files. The new approach intentionally removes silent degradation on format drift; cross-file callers from unchanged files are no longer part of the blast-radius graph. This is consistent with the no-second-parse-pass constraint and removes the text-format round-trip.

## Verification

- `cargo test --features "graph ast-context"`: 616 tests, 0 failed
- `cargo clippy --features "graph ast-context" -- -D warnings`: clean
- `cargo fmt --check`: clean
- `cargo check -p aptu-core --target wasm32-unknown-unknown --no-default-features`: 0 errors

Closes #1443

### CODE REVIEW

File: crates/aptu-core/src/ai/review_context.rs [hunk 1]
```diff
@@ -637,27 +651,132 @@ async fn build_ctx_graph(
...
+#[cfg(not(feature = "ast-context"))]
+async fn build_ctx_graph(
+    graph_config: &crate::config::GraphConfig,
+    repo_path: Option<&str>,
+    pr: &PrDetails,
+    _ast_text: &str,
+) -> (String, bool) {
+    let _ = (graph_config, repo_path, pr);
+    (String::new(), false)
+}
```
[the remark this answers is not in the document]
Comment (clouatre) [reply crates/aptu-core/src/ai/review_context.rs:687]: Replaced the chained `strip_prefix`/`or_else` logic with a `LazyLock<Regex>` pattern (regex crate is already a direct dep). The regex handles `async fn`, `pub(crate)`/`pub(super)` visibility modifiers, generic type parameters (`fn foo<T: Debug>`), tuple/unit structs, and `impl Trait for Type` (emitting only the implementing type name). Four new edge-case unit tests cover each of those patterns.

Regarding using `aptu-coder-core`'s parser (`analyze_str` on `PrFile.full_content`): that would be the ideal long-term direction and aligns with the recent move away from hand-rolled text parsers. The blocker for this iteration is that `full_content` is only populated for up to ~10 files per review and `analyze_str` is CPU-bound tree-sitter parsing that would need `spawn_blocking` inside the async call chain. Opened a follow-up to track that properly.

File: crates/aptu-core/src/ai/review_context.rs [hunk 2]
```diff
@@ -10,6 +10,23 @@ use std::path::PathBuf;
...
+use regex::Regex;
+#[cfg(all(feature = "ast-context", feature = "graph"))]
+use std::sync::LazyLock;
+
+/// Regex to extract symbol names from unified-diff added lines.
+/// Matches `fn`/`async fn`, `struct`, `enum`, `trait`, and `impl` declarations,
+/// stripping any visibility prefix (including `pub(crate)`, `pub(super)`, etc.).
+/// Capture group 2 is the keyword, capture group 3 is the symbol name.
+#[cfg(all(feature = "ast-context", feature = "graph"))]
+static SYMBOL_RE: LazyLock<Regex> = LazyLock::new(|| {
```
[the remark this answers is not in the document]
Comment (clouatre) [reply crates/aptu-core/src/ai/review_context.rs:23]: The regex is intentionally Rust-specific. The PR description's supported language list (Rust, Go, Python, TS, JS, C/C++, C#, Java) refers to the GitHub Contents API AST injection path, not this function. `derive_modified_symbols` only runs on Rust source files changed in the PR diff — the `SYMBOL_RE` pattern targets Rust declaration syntax (`fn`, `struct`, `enum`, `trait`, `impl`). Extending to Go or C# would require separate patterns and is a distinct concern; tracked in the follow-up referenced in the aptu-coder-core comment thread. No change needed here.

File: crates/aptu-core/src/graph/builder.rs [hunk 3]
```diff
@@ -1,252 +1,194 @@
...
-            let fn_idx = graph.add_node(Node::Function {
-                name: clean_name.clone(),
-                path: current_file_path.clone(),
-                visibility,
-            });
-            if let Some(file_idx) = current_file {
-                graph.add_edge(file_idx, fn_idx, Edge::Contains);
-            }
-            continue;
-        }
```
[the remark this answers is not in the document]
Comment (clouatre) [reply crates/aptu-core/src/graph/builder.rs:105]: Agreed — this is intentional. An empty `PathBuf` serializes to `"."" via `to_string_lossy`, which is meaningless as a caller path in the graph. Normalizing both `""` and `"."` to `String::new()` keeps the graph representation consistent with how anonymous callers are represented elsewhere.
